### ISSUE TITLE
JS Custom takes too long to set syntax on views

### ISSUE THREAD

**Issue by kylebebak** [Thom1729/Sublime-JS-Custom#58]
Based on an [issue that was originally created here](https://github.com/SublimeTextIssues/Core/issues/2603), but which probably belongs in this repo instead.

>I don't have a `JS Custom.sublime-settings` in my `User` directory -- I'm using the default settings. I've noticed this slowness with syntax highlighting of JS files for at least a few months.
>
>Here's a video of a different project, using the Goto File overlay. I'm holding the up arrow key and ST is previewing a bunch of JS (React) files, none of which is more than 500 lines long.
>
>https://streamable.com/hw1hb
>
>There's noticeable lag for setting syntax on most of the files, even ones with 150 LOC.

**Comment by kylebebak** [comment]
@Thom1729 

Any idea yet as to what's causing this? As a test, you could clone https://github.com/erikras/redux-form and try opening different files, or hitting **command+p** to preview different files.

When I do this I experience noticeable slowdown for any file with 200+ lines. I'm wondering if you experience the same thing.

**Comment by Thom1729** [comment]
I've been looking into this a bit. My usual benchmark is the [uncompressed JQuery source](https://code.jquery.com/jquery-3.3.1.js). For this file, the performance of my JS Custom – React configuration is similar to the core JavaScript syntax, and roughly in line with the [figures I obtained when rewriting that syntax](https://github.com/sublimehq/Packages/pull/1009#issuecomment-337435426). (Today, it took about 80 ms to parse all 10,364 lines.)

Of course, the jQuery source misses a lot of the content of the syntax. But most projects that use modern code (classes, JSX, etc) split the code across many files, making them more difficult to benchmark.

Are there any particular files that demonstrate the issue?

What platform are you on? I don't expect that it should matter, but it can't hurt.

Have you experienced the problem on 3.2? Sublime's Oniguruma version was recently updated. (This also shouldn't matter because the JavaScript syntax shouldn't be using Oniguruma anyway.)

**Comment by kylebebak** [comment]
Hey Thom, thanks a lot for trying to track down the source of the problem.

I'm on OSX 10.12.6, ST 3197 (latest version available), JS Custom 2.0.1.

I've done some debugging and benchmarking as well.

I uninstalled JS Custom, took the first 846 lines of jQuery (22998 chars) and made a file. Setting syntax on the file easily takes more than 0.5s. The slowdown is very obvious.

I then compared this with a [big Python file from youtube-dl](https://github.com/ytdl-org/youtube-dl/blob/master/youtube_dl/YoutubeDL.py), 2405 lines and 110576 chars. It sets syntax on the file very quickly probably ~0.1s.

In other words, per character, setting the regular JavaScript syntax on a file is **at least** 10x slower than setting Python syntax on a file.

I found similar speed for setting syntax on Go files.

So, it seems like it isn't a problem with JS Custom. It's a problem with the built-in JavaScript syntax.

This may mean it's a more difficult issue to solve. Still, I hope this is helpful -- at least now we know where to look.

**Comment by kylebebak** [comment]
Ok, useful update.

I uninstalled all my packages and now setting JS syntax on a view is blazing fast.

This is something that's definitely annoying about Sublime Text -- poorly written packages have undocumented behavior that can really fuck up the application =/

I'm sorry for sending you on this goose chase. When I figure out which package is the culprit I'll let you know.

**Comment by kylebebak** [comment]
Here it is:

<https://github.com/Monnoroch/ColorHighlighter>

**Comment by Thom1729** [comment]
All's well that ends well.

Related: https://github.com/SublimeTextIssues/Core/issues/2603
